File: GitToolBox/src/test/java/zielu/gittoolbox/tag/TagsLineListenerTest.java
package zielu.gittoolbox.tag;

import static org.assertj.core.api.Assertions.assertThat;

import com.intellij.execution.process.ProcessOutputTypes;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.stream.Stream;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

class TagsLineListenerTest {
  private TagsLineListener listener = new TagsLineListener();

  @DisplayName("Extract")
  @ParameterizedTest(name = "tags={1} from ''{0}''")
  @MethodSource("linesAndExpectedTags")
  void shouldExtractExpectedTags(String line, Collection<String> expectedTags) {
    listener.onLineAvailable(line, ProcessOutputTypes.STDOUT);
    assertThat(listener.getTags()).containsOnlyElementsOf(expectedTags);
  }

  private static Stream<Arguments> linesAndExpectedTags() {
    return Stream.of(
        Arguments.of("(HEAD -> develop)", Collections.emptyList()),
        Arguments.of("(HEAD -> master, tag: 1.2.2, origin/master, origin/HEAD)", Collections.singleton("1.2.2")),
        Arguments.of("(tag: 173.2.2, origin/master, master)", Collections.singleton("173.2.2")),
        Arguments.of("(origin/master, master, tag: 173.2.1)", Collections.singleton("173.2.1")),
        Arguments.of("(origin/develop)", Collections.emptyList()),
        Arguments.of("(tag: 173.1.2)", Collections.singleton("173.1.2")),
        Arguments.of("(tag: 173.1.2, tag: 173.2.2)", Arrays.asList("173.1.2", "173.2.2")),
        Arguments.of("(origin/172, 172)", Collections.emptyList()),
        Arguments.of("", Collections.emptyList()),
        Arguments.of("  ", Collections.emptyList())
    );
  }
}

File: src/test/kotlin/zielu/gittoolbox/tag/TagsLineListenerTest.kt
package zielu.gittoolbox.tag

import com.intellij.execution.process.ProcessOutputTypes
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

internal class TagsLineListenerTest {
  private val listener = TagsLineListener()

  @DisplayName("Extract")
  @ParameterizedTest(name = "tags={1} from ''{0}''")
  @MethodSource("linesAndExpectedTags")
  fun `should extract expected tags`(line: String?, expectedTags: Collection<String>) {
    listener.onLineAvailable(line, ProcessOutputTypes.STDOUT)
    assertThat(listener.tags).containsExactlyInAnyOrderElementsOf(expectedTags)
  }

  companion object {
    @JvmStatic
    private fun linesAndExpectedTags(): Stream<Arguments> {
      return Stream.of(
        Arguments.of("(HEAD -> develop)", emptyList<String>()),
        Arguments.of("(HEAD -> master, tag: 1.2.2, origin/master, origin/HEAD)", setOf("1.2.2")),
        Arguments.of("(tag: 173.2.2, origin/master, master)", setOf("173.2.2")),
        Arguments.of("(origin/master, master, tag: 173.2.1)", setOf("173.2.1")),
        Arguments.of("(origin/develop)", emptyList<String>()),
        Arguments.of("(tag: 173.1.2)", setOf("173.1.2")),
        Arguments.of("(tag: 173.1.2, tag: 173.2.2)", listOf("173.1.2", "173.2.2")),
        Arguments.of("(origin/172, 172)", emptyList<String>()),
        Arguments.of("", emptyList<String>()),
        Arguments.of("  ", emptyList<String>())
      )
    }
  }
}
